test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,282 @@
+                verbose=False,
+            )
+            return cpp_module
+
+        def serialize_module_into_file(module, example_inputs):
+            # We use JIT tracing to serialize Python module state, so that we can load it into C++
+            traced_script_module = torch.jit.trace(module, example_inputs)
+            module_file = tempfile.NamedTemporaryFile(delete=False)
+            traced_script_module.save(module_file.name)
+            module_file.close()
+            return module_file.name
+
+        def test_method(method_name, test_params, cpp_module):
+            device = test_params.device
+            input_size = test_params.input_size
+            input_fn = test_params.input_fn
+            python_module_class = test_params.python_module_class
+            python_constructor_args = test_params.python_constructor_args
+            module_variant_name = test_params.module_variant_name
+            cpp_test_name = module_variant_name + '_test_' + method_name
+
+            if input_size:
+                example_inputs = [torch.randn(input_size)]
+            elif input_fn:
+                example_inputs = list(input_fn())
+            else:
+                raise RuntimeError("Missing `input_size` or `input_fn` for {}".format(module_variant_name))
+            example_inputs = [x.to(device) for x in example_inputs]
+
+            if method_name == 'init':
+                args = test_init(device, python_module_class, python_constructor_args)
+            elif method_name == 'forward':
+                args = test_forward(device, python_module_class, python_constructor_args, example_inputs)
+            elif method_name == 'backward':
+                args = test_backward(device, python_module_class, python_constructor_args, example_inputs)
+            else:
+                raise RuntimeError("{} is not a supported method to test".format(method_name))
+
+            modules = args[0]
+            module_file_names = []
+            for module in modules:
+                module_file_name = serialize_module_into_file(module, example_inputs)
+                module_file_names.append(module_file_name)
+
+            cpp_args = []
+            for module_file_name in module_file_names:
+                cpp_args.append(module_file_name)
+            for arg in args[1:]:
+                if isinstance(arg, list):
+                    cpp_args += arg
+                else:
+                    cpp_args.append(arg)
+            try:
+                getattr(cpp_module, cpp_test_name)(*cpp_args)
+            finally:
+                # Ideally we would like to not have to manually delete the file, but NamedTemporaryFile
+                # opens the file, and it cannot be opened multiple times in Windows. To support Windows,
+                # we close the file after creation and try to remove it manually.
+                for module_file_name in module_file_names:
+                    os.unlink(module_file_name)
```

Comment:
Can this fail? Maybe we should catch and send failures to stderr so that the test still passes (maybe with a warning?)